### PR TITLE
Semi specific digestion

### PR DESCRIPTION
**Pull request by Jeremi-Maciejewski**
## Description

These changes implement support of performing semi-specific digestion to ProteaseDigestion, as requested in #8098 .
Also exposed via pyOpenMS.

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

## Summary by CodeRabbit

* **New Features**
  * Added support for semi-specific digestion in protease digestion, allowing generation of peptides with only one end at a cleavage site when specificity is set accordingly.

* **Documentation**
  * Updated user documentation and code comments to describe semi-specific digestion functionality and provide usage examples.

* **Tests**
  * Introduced new unit tests to verify semi-specific digestion and related behaviors.
  * Removed outdated tests for protease digestion.
  * Added test coverage for error handling and edge cases in semi-specific digestion.

* **Chores**
  * Added a new contributor to the authors list.
  * Updated digestion specificity in decoy peptide generation to enforce full enzymatic specificity.

### CODE REVIEW

File: src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h [hunk 1]
```diff
@@ -179,6 +179,21 @@ namespace OpenMS
      */
     std::vector<int> tokenize_(const String& sequence, int start = 0, int end = -1) const;
 
+    /**
+      @brief Generates semi-specific digestion products
+
+      Function handling semi-specific digestion (specificity == 1).
```
Comment (cbielow) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h]:
```suggestion
      Function handling semi-specific digestion (specificity_ == Specificity::SPEC_SEMI ).
```

File: src/openms/include/OpenMS/CHEMISTRY/ProteaseDigestion.h [hunk 2]
```diff
@@ -49,7 +49,9 @@ namespace OpenMS
     Size digest(const AASequence& protein, std::vector<AASequence>& output, Size min_length = 1, Size max_length = 0) const;
 
     /** 
-       @brief Performs the enzymatic digestion of a protein represented as AASequence
+       @brief Performs the enzymatic digestion of a protein represented as AASequence.
+       By default, only fully specific products are generated.
+       If object's Specificity is set to SPEC_SEMI (1), also generates semi-specific products.
```
Comment (cbielow) [inline src/openms/include/OpenMS/CHEMISTRY/ProteaseDigestion.h]: There is no `default` here (when function parameters are concerned).
-->
```
Digestion takes into account the value of
 - the **missed cleavages** member, see @p setMissedCleavages()
 - the **specificity** member, see @p setSpecificity()

Currently, Specificity::SPEC_SEMI and Specificity::SPEC_FULL are supported. Others raise an exception

@throw Exception::InvalidValue   // <-- todo

```

```

Comment (cbielow) [reply src/openms/include/OpenMS/CHEMISTRY/ProteaseDigestion.h]: please also update the docs of `Size digest(const AASequence& protein, ....` since that is just a wrapper.
(maybe just add a doxgen reference pointing here for details)

File: src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h [hunk 3]
```diff
@@ -179,6 +179,21 @@ namespace OpenMS
      */
     std::vector<int> tokenize_(const String& sequence, int start = 0, int end = -1) const;
 
+    /**
+      @brief Generates semi-specific digestion products
```
Comment (cbielow) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h:183]: will it skip fully specific variants?

will it take MC's into account?

Please document

File: src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp [hunk 4]
```diff
@@ -115,6 +115,65 @@ namespace OpenMS
...
+    const int last_c = cleavage_positions.back(); // Last cleavage
+    const int lgth = last_c - first_c;
+
+    // Lambda checking min & max conditions and adding to output
+    auto variant = [&output, &wrong, min_length, max_length](Size x, Size y)
+    {
+      if (min_length <= std::abs((int)y - (int)x) &&
+          max_length >= std::abs((int)y - (int)x))
+      {
+        output.emplace_back(x, y);
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:154]: since x < y by this definition, do you need std::abs above?

File: src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp [hunk 5]
```diff
@@ -115,6 +115,65 @@ namespace OpenMS
...
+      } else ++wrong;
+    };
+
+    auto fwd = cleavage_positions.cbegin();
+    auto bwd = cleavage_positions.crbegin(); // Reverse iterator!
+    ++fwd; ++bwd;
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:160]: dangerous!
check `cleavage_positions.empty()` and return as a first action in the function.

File: src/openms/source/CHEMISTRY/ProteaseDigestion.cpp [hunk 6]
```diff
@@ -139,6 +139,13 @@ namespace OpenMS
         }
       }
     }
+
+    // semi-specific variants
+    if (specificity_ == SPEC_SEMI)
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/ProteaseDigestion.cpp:152]: test for specificity_  and throw exception unless its SEMI or FULL

File: src/pyOpenMS/tests/unittests/test_ProteaseDigestion.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    seq = AASequence.fromString("MHARLVP")
+    output = []
+    ff.digest(seq, output)
+    # MHAR, LVP
+    assert len(output) == 2
+    ff.setSpecificity(1) # Semi-specific
+    output = []
+    ff.digest(seq, output)
+    # MHAR, LVP, HAR, LV, AR, L, R, MHA, VP, MH, P, M
+    assert len(output) == 12
```
Comment (cbielow) [inline src/pyOpenMS/tests/unittests/test_ProteaseDigestion.py:48]: test the new exception for e.g., Specificity UNKNOWN

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp [hunk 8]
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
...
     TEST_EQUAL(out.size(), 4*3/2)
+
+    // ------------------------
+    // semi-specific digestion
+    // ------------------------
+    pd.setSpecificity(pd.SPEC_SEMI);
+
+    // Lambda to allow searching 'out' with std::find_if()
+    std::string ref = "";
+    auto compare_aa = [&ref](AASequence seq) -> bool
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]:
```suggestion
    auto compare_aa = [&ref](const AASequence& seq) -> bool
```

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp [hunk 9]
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
...
+    pd.digest(AASequence::fromString("AGLRMHP"), out);
+    // AGLR, MHP, GLR, MH, LR, M, R, AGL, HP, AG, P, A
+    TEST_EQUAL(out.size(), 12);
+    // Assumes output is sorted as: fully-specific -> missed clv -> semi-specific
+    // Are fully-specific variants still here?
+    TEST_EQUAL(out[0].toString(), "AGLR");
+    TEST_EQUAL(out[1].toString(), "MHP");
+    // Are there semi-specific variants?
+    ref = "AGL";
+    TEST_EQUAL(std::find_if(out.begin(),out.end(),compare_aa) != out.end(), true);
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: you could use TEST_TRUE here

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp [hunk 10]
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
...
+    ref = "HP";
+    TEST_EQUAL(std::find_if(out.begin(),out.end(),compare_aa) != out.end(), true);
+    // Make sure cleavage sites are iterated over in right order.
+    // (forward for N-terminus cleavage, backwards for C-terminus cleavage)
+    ref = "GLRMHP";
+    TEST_NOT_EQUAL(std::find_if(out.begin(),out.end(),compare_aa) != out.end(), true);
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: TEST_FALSE? :)

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp [hunk 11]
```diff
@@ -227,6 +228,66 @@ START_SECTION((Size digest(const AASequence& protein, std::vector<AASequence>& o
...
+
+    // Make sure unspecific cleavage still works
+    pd.setEnzyme("unspecific cleavage");
+    pd.digest(AASequence::fromString("ABCDEFGHIJ"), out);
+    TEST_EQUAL(out.size(), 11*10/2);
+    pd.digest(AASequence::fromString("ABC"), out);
+    TEST_EQUAL(out.size(), 4*3/2);
+
+    pd.setEnzyme("Trypsin/P");
+    pd.digest(AASequence::fromString("AGLRMHP"), out);
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp:252]: can you test this with a protein where the MC setting is less than the actual number of more cleavage sites?! Just to make sure this is done correctly.
Also, the other way around: A protein with no cleavage sites and MC=1 should produce valid output.

**Review by cbielow — CHANGES_REQUESTED** [review]
nice addition, thanks!

some comments below

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp (hunk 9 above)
Comment (Jeremi-Maciejewski) [reply src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: I used TEST_EQUAL since its documentation says:
`If possible use TEST_EQUAL(a, b) instead of TEST_TRUE(a==b), because the latter makes bug tracing harder.`

(same for TEST_FALSE and TEST_NOT_EQUAL)

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp (hunk 8 above)
Comment (Jeremi-Maciejewski) [reply src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: Corrected.

File: src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp (hunk 4 above)
Comment (Jeremi-Maciejewski) [reply src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:154]: Not really, I removed it.

File: src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h (hunk 3 above)
Comment (Jeremi-Maciejewski) [reply src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h:183]: Done.

File: src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h (hunk 1 above)
Comment (Jeremi-Maciejewski) [reply src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h]: Corrected.

File: src/openms/source/CHEMISTRY/ProteaseDigestion.cpp (hunk 6 above)
Comment (Jeremi-Maciejewski) [reply src/openms/source/CHEMISTRY/ProteaseDigestion.cpp:152]: Done.

File: src/pyOpenMS/tests/unittests/test_ProteaseDigestion.py (hunk 7 above)
Comment (Jeremi-Maciejewski) [reply src/pyOpenMS/tests/unittests/test_ProteaseDigestion.py:48]: Done. (although with SIZE_OF_SPECIFICITY)

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp (hunk 9 above)
Comment (cbielow) [reply src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: we should update the docs, sorry for the confusion.
it should say, `if you test for a boolean, use TEST_TRUE / TEST_FALSE, otherwise prefer TEST_EQUAL/TEST_NOT_EQUAL to ease bug tracing.`
E.g. do not do this
`TEST_EQUAL(myStringFunc() == "some_return_value", true)`, because it will fail, but not say what the actual return value was

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp (hunk 11 above)
Comment (Jeremi-Maciejewski) [reply src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp:252]: First was already implemented, I added test for 2nd case.

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp (hunk 9 above)
Comment (Jeremi-Maciejewski) [reply src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: I see, that does make more sense. Thanks for explaining, I'll change it to `TEST_TRUE`/`TEST_FALSE` .

File: src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp (hunk 5 above)
Comment (Jeremi-Maciejewski) [reply src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:160]: Implemented (throws `Exception::InvalidValue` on too few cleavage positions - should actually be at least 2)

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp (hunk 9 above)
Comment (Jeremi-Maciejewski) [reply src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: Corrected

File: src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp (hunk 10 above)
Comment (Jeremi-Maciejewski) [reply src/tests/class_tests/openms/source/ProteaseDigestion_test.cpp]: Corrected.

File: src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp [hunk 12]
```diff
@@ -115,6 +115,73 @@ namespace OpenMS
...
   }
 
+  Size EnzymaticDigestion::semiSpecificDigestion_(const std::vector<int>& cleavage_positions, std::vector<std::pair<Size, Size>>& output, Size min_length, Size max_length) const
+  {
+    // Too few cleavage sites - should be at least sequence start and end
+    if (cleavage_positions.size() < 2)
+    {
+      String value(cleavage_positions.begin(), cleavage_positions.end());
+      throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION,
+        String("Too few cleavage positions - at least sequence start and end positions are required."), value);
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:125]: document that using Doxygen

File: src/tests/class_tests/openms/source/EnzymaticDigestion_test.cpp [hunk 13]
```diff
@@ -282,6 +282,33 @@ START_SECTION((Size digestUnmodified(const StringView sequence, std::vector<Stri
...
 
+START_SECTION((Size semiSpecificDigestion_(const std::vector<int>& cleavage_positions, std::vector<std::pair<Size, Size>>& output, Size min_length, Size max_length) const))
+{
+    class TempChild : public EnzymaticDigestion
+    {
+        public:
+            void tmpSetEnzyme(const DigestionEnzyme* enzyme)
+            {
+                this->setEnzyme(enzyme);
+            }
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/EnzymaticDigestion_test.cpp]: do you need this overload?
`setEnyme()` is public.

File: CHANGELOG [hunk 14]
```diff
@@ -29,6 +29,9 @@ Dependencies:
 NucleicAcidSearchEngine:
     - add support for global fixed modifications as a search parameter
 
+ProteaseDigestion:
+    - add support for semi-specific digestion in objects where specificity is set to SPEC_SEMI (1)
```
Comment (cbielow) [inline CHANGELOG]: move this to  Library and mention PR number:
```
Library:
 ...
  - ProteaseDigestion: add support for semi-specific digestion (#8130)
```

**Review by cbielow — COMMENTED** [review]
nice. almost there :)

File: src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp (hunk 12 above)
Comment (Jeremi-Maciejewski) [reply src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp:125]: Done.

File: src/tests/class_tests/openms/source/EnzymaticDigestion_test.cpp (hunk 13 above)
Comment (Jeremi-Maciejewski) [reply src/tests/class_tests/openms/source/EnzymaticDigestion_test.cpp]: Right, I do not. Removed.

File: CHANGELOG (hunk 14 above)
Comment (Jeremi-Maciejewski) [reply CHANGELOG]: Done.
